kernel/devfs.c

```diff
@@ -81,6 +99,134 @@ static int _urandom_read_cb(void* buf, size_t count)
     return ret;
 }
 
+static struct pty_pair* _find_paired_master_by_path(const char* path)
+{
+    for (struct pty_pair* tmp = _pty_pairs; tmp; tmp = tmp->next)
+    {
+        if (strcmp(path, tmp->path_slave) == 0)
+            return tmp;
+    }
+    return NULL;
+}
+
+static int _open_slave_pty_cb(
+    myst_file_t* file,
+    myst_buf_t* buf,
+    const char* path)
+{
+    (void)buf;
+    int ret = 0;
+    // find the paired master based on path
+    struct pty_pair* pair = _find_paired_master_by_path(path);
+    if (pair == NULL)
+        ERAISE(-ENXIO);
+    pair->file_slave = file;
+
+done:
+    return ret;
+}
+
+static int _close_pty_file_cb(myst_file_t* file)
+{
+    int ret = 0;
+    // free the pair
+    struct pty_pair* prev = NULL;
+    for (struct pty_pair* tmp = _pty_pairs; tmp; tmp = tmp->next)
+    {
+        if (tmp->file_master == file || tmp->file_slave == file)
+        {
+            if (tmp == _pty_pairs)
+                prev->next = tmp->next;
+            else
+                _pty_pairs = _pty_pairs->next;
+
+            free(tmp->path_master);
+            free(tmp->path_slave);
+            free(tmp);
+        }
+        prev = tmp;
+    }
+
+    return ret;
+}
+
+static int _read_master_pty_cb(myst_file_t* file, void* buf, size_t count)
+{
+    int ret = 0;
+    for (struct pty_pair* tmp = _pty_pairs; tmp; tmp = tmp->next)
+    {
+        if (tmp->file_master == file)
+        {
+            ret = myst_read_virtual_file(tmp->file_slave, buf, count);
+            goto done;
+        }
+    }
+
+    ERAISE(-EINVAL); /* can't find the paired slave PTY device */
+
+done:
+    return ret;
+}
+
+static int _read_slave_pty_cb(myst_file_t* file, void* buf, size_t count)
+{
+    int ret = 0;
+    for (struct pty_pair* tmp = _pty_pairs; tmp; tmp = tmp->next)
+    {
+        if (tmp->file_slave == file)
+        {
+            ret = myst_read_virtual_file(tmp->file_master, buf, count);
+            goto done;
+        }
+    }
+    ERAISE(-EINVAL); /* can't find the paired slave PTY device */
+
+done:
+    return ret;
+}
+
+static int _write_pty_cb(myst_file_t* file, const void* buf, size_t count)
+{
+    return myst_write_virtual_file(file, buf, count);
+}
+
+static int _open_master_pty_cb(
+    myst_file_t* file,
+    myst_buf_t* buf,
+    const char* path)
+{
+    (void)buf;
+    int ret = 0;
+    char tmp[64];
+    myst_vcallback_t v_cb = {0};
+    struct pty_pair* pair = calloc(1, sizeof(struct pty_pair));
+    if (pair == NULL)
+        ERAISE(-ENOMEM);
+
+    if ((pair->path_master = strdup(path)) == NULL)
+        ERAISE(-ENOMEM);
+
+    pair->slaveID = _nextSlaveID++;
+    snprintf(tmp, sizeof(tmp), "/pts/%d", pair->slaveID);
+
+    if ((pair->path_slave = strdup(tmp)) == NULL)
+        ERAISE(-ENOMEM);
+
+    pair->file_master = file;
+    pair->next = _pty_pairs;
+    _pty_pairs = pair;
+
+    v_cb.open_cb = _open_slave_pty_cb;
+    v_cb.close_cb = _close_pty_file_cb;
```

Comment:
When is `close_cb` used?

Reply:
It used by `_close_pty_file_cb`, where we tear down the master/slave pair when either the master or the slave device is closed.

Reply:
Gotcha. So how is that function being invoked from userspace. Is it called when the file handle is closed?

Reply:
Thanks for catching that. Added the plumbing in _fs_close so the call back can be triggered.